### PR TITLE
Add library boilerplate and strict(-er) mocking functionality

### PR DESCRIPTION
**Pull request by ben-challis**
Adds initial functionality to override and disable return value generation to reduce the risk of refactoring where new method expectations are not set and default values may result in a passing test suite.

### CODE REVIEW

File: .gitattributes [hunk 1]
```diff
@@ -0,0 +1,10 @@
+/.github            export-ignore
+/tests              export-ignore
```
Comment (marmichalski) [inline .gitattributes:2]:
```suggestion
/tests              export-ignore
/tmp                export-ignore
```

File: .gitattributes [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+.gitattributes      export-ignore
```
Comment (marmichalski) [inline .gitattributes:3]:
```suggestion
.gitattributes      export-ignore
.allowed-licenses.php export-ignore
```

will need indentation adjusting 😭

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+on:
+  merge_group:
+  pull_request:
+  push:
+    branches:
+      - "main"
+
+jobs:
+  ci:
+    name: "Execute (${{ matrix.php-version }} @ ${{ matrix.dependencies }})"
```
Comment (marmichalski) [inline .github/workflows/ci.yaml:16]:
```suggestion
    name: "Execute (PHP ${{ matrix.php-version }} @ ${{ matrix.dependencies }} deps)"
```

File: .github/workflows/ci.yaml [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+          - "highest"
+          - "locked"
+        php-version:
+          - "8.1"
+          - "8.2"
+          - "8.3"
+
+    steps:
+      - name: "Checkout"
+        uses: "actions/checkout@v4"
```
Comment (marmichalski) [inline .github/workflows/ci.yaml:33]:
```suggestion
        uses: "actions/checkout@v4"
        with:
          show-progress: false
```

File: .github/workflows/ci.yaml [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+          dependency-versions: "${{ matrix.dependencies }}"
+
+      - name: "Validate Composer dependencies"
+        run: "composer validate"
+
+      - name: "Check Composer dependency licenses"
+        run: "composer licenses:check"
+
+      - name: "Run Composer audit"
+        run: "composer audit"
```
Comment (marmichalski) [inline .github/workflows/ci.yaml:65]:
```suggestion
```

I think we don't need this here? Or do we?

File: phpunit.xml.dist [hunk 6]
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:noNamespaceSchemaLocation="vendor/phpunit/phpunit/phpunit.xsd"
+         colors="true"
+         cacheResult="false"
+         cacheDirectory="tmp/phpunit"
+         bootstrap="vendor/autoload.php">
+    <source>
+        <include>
+            <directory>./lib/</directory>
```
Comment (ben-challis) [inline phpunit.xml.dist:10]:
```suggestion
            <directory>./src/</directory>
```

File: phpstan-rector.neon [hunk 7]
```diff
@@ -0,0 +1,8 @@
+includes:
+  - vendor/phpstan/phpstan-phpunit/extension.neon
+
+parameters:
+    paths:
+      - src
+      - tests
+    level: max
```
Comment (marmichalski) [inline phpstan-rector.neon:8]:
```suggestion
  paths:
    - src
    - tests
  level: max
```

File: src/StrictMocking.php [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+use PHPUnit\Framework\MockObject\Generator\Generator as MockObjectGenerator;
+use PHPUnit\Framework\MockObject\MockObject;
+use PHPUnit\Framework\TestCase;
+
+/**
+ * @mixin TestCase
+ */
+trait StrictMocking
+{
+    protected function createMock(string $originalClassName): MockObject
```
Comment (ben-challis) [inline src/StrictMocking.php:17]:
```suggestion
    final protected function createMock(string $originalClassName): MockObject
```

Doesn't change much but intent.

File: phpstan.neon [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+  - vendor/phpstan/phpstan-strict-rules/rules.neon
+  - phar://vendor/phpstan/phpstan/phpstan.phar/conf/bleedingEdge.neon
+
+parameters:
+    tmpDir: tmp/phpstan
+    paths:
+      - src
+      - tests
+    level: max
+    checkExplicitMixed: true
```
Comment (marmichalski) [inline phpstan.neon:13]:
```suggestion
  tmpDir: tmp/phpstan
  paths:
    - src
    - tests
  level: max
  checkExplicitMixed: true
```

File: phpunit.xml.dist (hunk 6 above)
Comment (marmichalski) [inline phpunit.xml.dist]:
```suggestion
            <directory>./src/</directory>
```

**Review by marmichalski — APPROVED** [review]
👌

**Review by marmichalski — APPROVED** [review]

**Review by marmichalski — APPROVED** [review]

**Pull request merged**